pkg/util/kubernetes/hpa/hpa.go

```diff
@@ -0,0 +1,361 @@
+type HPAWatcherClient struct {
+	apiClient         *as.APIClient
+	hpaReadTimeout    time.Duration
+	hpaRefreshItl     time.Duration
+	hpaPollItl        time.Duration
+	hpaExternalMaxAge time.Duration
+}
+
+func (c *HPAWatcherClient) hpaWatcher(res string) (new []*v2beta1.HorizontalPodAutoscaler, modified []*v2beta1.HorizontalPodAutoscaler, deleted []*v2beta1.HorizontalPodAutoscaler, resVer string, err error) {
+	hpaInterface := c.apiClient.ClientSet.AutoscalingV2beta1()
+	metaOptions := metav1.ListOptions{Watch: true, ResourceVersion: res}
+
+	hpaWatcher, err := hpaInterface.HorizontalPodAutoscalers(metav1.NamespaceAll).Watch(metaOptions)
+	if err != nil {
+		log.Infof("Failed to watch %v: %v", expectedHPAType, err)
+	}
+	defer hpaWatcher.Stop()
+
+	watcherTimeout := time.NewTimer(c.hpaReadTimeout)
+	for {
+		select {
+		case rcvdHPA, ok := <-hpaWatcher.ResultChan():
+			if !ok {
+				log.Debugf("Unexpected watch close")
+				return nil, nil, nil, "0", err
+			}
+			currHPA, ok := rcvdHPA.Object.(*v2beta1.HorizontalPodAutoscaler)
+			if !ok {
+				log.Infof("Wrong type: %s", currHPA)
+				continue
+			}
+			if currHPA.ResourceVersion != "" && currHPA.ResourceVersion != resVer {
+				resVer = currHPA.ResourceVersion
+			}
+			if rcvdHPA.Type == watch.Error {
+				log.Infof("Error in the watcher, evaluating: %s", currHPA)
+				//status, ok := rcvdHPA.Object.()
+			}
+			if rcvdHPA.Type == watch.Added {
+				log.Infof("Adding this manifest: %s", currHPA)
+				new = append(new, currHPA)
+			}
+			if rcvdHPA.Type == watch.Modified {
+				log.Infof("Modifying this manifest: %s", currHPA)
+				modified = append(modified, currHPA)
+			}
+			if rcvdHPA.Type == watch.Deleted {
+				deleted = append(deleted, currHPA)
+			}
+
+			watcherTimeout.Reset(c.hpaReadTimeout)
+		case <-watcherTimeout.C:
+			return new, modified, deleted, resVer, nil
+		}
+	}
+}
+
+func newHPAWatcher() *HPAWatcherClient {
+	clientAPI, err := as.GetAPIClient()
+	if err != nil {
+		log.Errorf("Error creating Client for the HPA: %s", err.Error())
+		return nil
+	}
+	hpaPollItl := config.Datadog.GetInt("hpa_watcher_polling_freq")
+	hpaRefreshItl := config.Datadog.GetInt("hpa_external_metrics_polling_freq")
+	hpaExternalMaxAge := config.Datadog.GetInt("hpa_external_metrics_max_age")
+	return &HPAWatcherClient{
+		apiClient:         clientAPI,
+		hpaReadTimeout:    100 * time.Millisecond,
+		hpaPollItl:        time.Duration(hpaPollItl) * time.Second,
+		hpaRefreshItl:     time.Duration(hpaRefreshItl) * time.Second,
+		hpaExternalMaxAge: time.Duration(hpaExternalMaxAge) * time.Second,
+	}
+}
+
+// GetHPAWatcherClient returns the HPAWatcherClient
+func GetHPAWatcherClient() *HPAWatcherClient {
+	return newHPAWatcher()
+}
+
+// HPAWatcher runs a watch process of the various HPA objects' activities to process and store the relevant info.
+// Refreshes the custom metrics stored as well.
+func (c *HPAWatcherClient) HPAWatcher() {
```

Comment:
Would it be possible to add a `Stop()` function to the `HPAWatcherClient` to cleanup all running goroutines?

Also, I think we can call this function simply `Start()` since the name of the type already communicates really well what will happen.